walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -35,106 +40,263 @@ public int compare(Event a, Event b) {
-      addIfLongEnough(
-          TimeRange.fromStartEnd(0, END_OF_DAY, true), possibleMeetingTimes, request.getDuration());
-      return possibleMeetingTimes;
+    /**
+     * Returns a list of time periods in which the meeting, specified by request, could happen. If
+     * no time exists for all optional and mandatory attendees, find the time slot(s) that allow
+     * mandatory attendees and the greatest possible number of optional attendees to attend.
+     */
+    Collection<TimeRange> resolveBestTime() {
+      ArrayList<TimeRange> mandatoryAttendeesMeetingTimes =
+          getMandatoryAttendeesMeetingTimes(
+              request.getDuration(), new HashSet<String>(request.getAttendees()));
+      Collection<TimeRange> optimalMeetingTimes =
+          getOptimalMeetingTimes(
+              mandatoryAttendeesMeetingTimes,
+              getChangesInOptionalAttendeesAttendance(
+                  getOptionalAttendeesFreeTimes(
+                      new HashSet<String>(request.getOptionalAttendees()), request.getDuration())),
+              request.getDuration());
+
+      // If there are no meeting times with at least one optional attendee, just return
+      // mandatoryAttendeesMeetingTimes.
+      return optimalMeetingTimes.isEmpty() ? mandatoryAttendeesMeetingTimes : optimalMeetingTimes;
     }
-    Collections.sort(events, ORDER_BY_START_ASC);
-
-    // Add first gap.
-    addIfLongEnough(
-        TimeRange.fromStartEnd(0, events.get(0).getWhen().start(), false),
-        possibleMeetingTimes,
-        request.getDuration());
-    int end = events.get(0).getWhen().end();
-    for (Event event : events) {
-      // event can be merged with current time range
-      if (event.getWhen().start() <= end) {
-        end = Math.max(end, event.getWhen().end());
-        continue;
+
+    /**
+     * Returns all meeting times that allow the most optional attendees to attend. These times must
+     * also fall in a time satisfying all mandatory attendees and must be at least minTime long.
+     * Utilizes two pointers: one to mandatoryAttendeesMeetingTimes, one to changeLog.
+     *
+     * @param mandatoryAttendeesMeetingTimes all meeting times satisfying mandatory attendees
+     * @param changeLog a change log of the number of available optional attendees over time
+     * @param minTime the minimum time a meeting must last for
+     */
+    private Collection<TimeRange> getOptimalMeetingTimes(
+        ArrayList<TimeRange> mandatoryAttendeesMeetingTimes,
+        TreeMap<Integer, Integer> changes,
+        long minTime) {
+
+      ArrayList<TimeRange> optimalMeetingTimes = new ArrayList<TimeRange>();
+      ArrayList<Map.Entry<Integer, Integer>> changeLog =
+          new ArrayList<Map.Entry<Integer, Integer>>(changes.entrySet());
+
+      // changeLogIndexis a pointer in changeLog,  mandatoryAttendeesMeetingTimesIndex is a pointer
+      // in mandatoryAttendeesMeetingTimes.
+      int mandatoryAttendeesMeetingTimesIndex = 0,
```

Comment:
As a general rule, indexes into collections should be the iteration method of last resort because algorithms using them are always harder to understand than algorithms that avoid them (i.e. by using for-each, or using a regular iterator).

Reply:
Got it, l think in this case it's probably easier to use an index because sometimes we need to go back one in `mandatoryAttendeesMeetingTimes`, so l'll keep it for this case.

Reply:
Ack, this deals with the comment two below this, so l'll resolve it.